### PR TITLE
refactor: React components from CMS layer & native react props

### PR DESCRIPTION
**Pull request by chohner**
Cleans up interface between CMS and React components (currentl for all form components)

- ~Native props for React components (no zod)~
- ~React components can handle `null` (if easily possible)~
- CMS translation layer exports React components (no `xFromStrapi()`, no `getXProps()` handling)
- ~Splitting CMS model into `XSchema` (required props) and `XComponentSchema` (with required `__component` prop used inside dynamic zones)~

### CODE REVIEW

File: app/components/Image.tsx [hunk 1]
```diff
@@ -2,14 +2,14 @@ import { z } from "zod";
 
 export const ImagePropsSchema = z
   .object({
-    url: z.string(),
+    url: z.string().nullish(),
```
Comment (pgurusinga) [inline app/components/Image.tsx:5]: should this be enforced? if there's no URL then the image would look like a broken link?

File: app/components/inputs/ErrorMessageProps.ts [hunk 2]
```diff
@@ -0,0 +1,4 @@
+export type ErrorMessageProps = {
```
Comment (pgurusinga) [inline app/components/inputs/ErrorMessageProps.ts:1]: why does this type live in the input directory? from the name of it, it is a pretty generic error message props

Comment (chohner) [reply app/components/inputs/ErrorMessageProps.ts:1]: Good question. It was `index.ts` before, lets find a better spot for it

File: app/components/Image.tsx (hunk 1 above)
Comment (chohner) [reply app/components/Image.tsx:5]: The component itself returns `null` without url. Not sure why, but I didnt want to change the functionality

Comment (pgurusinga) [reply app/components/Image.tsx:5]: ah ok, this image is not an independent component, so hard to follow the types.

File: app/services/cms/models/StrapiCheckbox.tsx [hunk 3]
```diff
@@ -16,15 +15,19 @@ export const StrapiCheckboxSchema = z
...
-export const renderCheckboxFromStrapi = (
-  strapiCheckbox: z.infer<typeof StrapiCheckboxSchema>,
-) => (
+type StrapiCheckbox = z.infer<typeof StrapiCheckboxSchema>;
+
+export const StrapiCheckboxComponentSchema = StrapiCheckboxSchema.extend({
+  __component: z.literal("form-elements.checkbox"),
+});
+
+export const StrapiCheckbox = ({
```
[the remark this answers is not in the document]
Comment (chohner) [reply app/services/cms/models/StrapiCheckbox.tsx:24]: Hm I couldn't reproduce this. I merged main into this just in case, could you please check again? 🙏🏼

[the remark this answers is not in the document]
Comment (SannyNguyenHung) [reply app/services/cms/models/StrapiCheckbox.tsx:24]: Its gone :shrug:

File: app/services/cms/models/StrapiContentComponent.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    StrapiBoxWithImageSchema.required({ __component: true }),
+    StrapiHeaderSchema.required({ __component: true }),
+    StrapiInfoBoxSchema.required({ __component: true }),
+    StrapiInfoBoxItemSchema.required({ __component: true }),
+    StrapiParagraphSchema.required({ __component: true }),
+    StrapiLinkListBoxSchema.required({ __component: true }),
+    StrapiListSchema.required({ __component: true }),
+    StrapiListItemSchema.required({ __component: true }),
+    StrapiAlertSchema.required({ __component: true }),
+    StrapiArraySummarySchema.required({ __component: true }),
```
[the remark this answers is not in the document]
Comment (chohner) [reply app/services/cms/models/StrapiContentComponent.ts:29]: I only moved the form components in this PR. If we agree on moving into any direction I would adapt the content components too.

File: app/components/PageContent.tsx [hunk 5]
```diff
@@ -75,47 +73,43 @@ function wrapInBackground(cmsData: StrapiContent, reactElement: ReactElement) {
...
     case "page.link-list-box":
-      return <LinkListBox {...getLinkListBoxProps(strapiContent)} key={key} />;
+      return <LinkListBox {...getLinkListBoxProps(strapiContent)} />;
     case "page.box-with-image":
-      return (
-        <BoxWithImage {...getBoxWithImageProps(strapiContent)} key={key} />
-      );
+      return <BoxWithImage {...getBoxWithImageProps(strapiContent)} />;
     case "page.list":
-      return <List {...getListProps(strapiContent)} key={key} />;
```
[the remark this answers is not in the document]
Comment (chohner) [reply app/components/PageContent.tsx:121]: The mapping over the list happens inside `PageComponent()`, where we have a `<div>` that takes the key. The wrapped elements don't need key (I think this was refactored and just forgotten). I'll also move this into its own commit

File: app/components/PageContent.tsx [hunk 6]
```diff
@@ -75,47 +73,43 @@ function wrapInBackground(cmsData: StrapiContent, reactElement: ReactElement) {
...
-      return <RichText {...getRichTextProps(strapiContent)} key={key} />;
+      return <RichText {...getRichTextProps(strapiContent)} />;
     case "basic.alert":
       return renderAlertFromStrapi(strapiContent);
     case "page.header":
-      return <Header {...getHeaderProps(strapiContent)} key={key} />;
+      return <Header {...getHeaderProps(strapiContent)} />;
     case "form-elements.input":
-      return <Input {...getInputProps(strapiContent)} key={key} />;
+      return <StrapiInput {...strapiContent} />;
```
[the remark this answers is not in the document]
Comment (chohner) [reply app/components/PageContent.tsx:86]: Simply because I only migrated the `StrapiFormComponents` to see how this pattern looks like

**Comment by chohner** [comment]
Thanks @frederike-ramin for your thoughtful review 🙏🏼 

I feel like there are two topics:

## `null` properties
`null` seems to be an uncommon value to think about/deal with in the front-end, but I haven't really found a specific best practice for this. This is a bit surprising to me because in many cases there is no implementation overhead of dealing with both `undefined` and `null` (something like `image && {image}` or `text ?? "default"` works for both). The only additional work is adding the type (aka `text?: string | null`).

I usually have a slight preference to be permissive with the inputs, especially if it comes at no implementation / runtime cost. If there is a general preference to remove `null` from props I'd be fine with that of course. I did it here because it simplified the mapping a little bit but we could also just call `omitNull()` a few more times 😄 

## Separation of models and view
### Current state
There is a single frontend component `<PageContent>` (which maybe should be called `<StrapiComponents>`?) that takes an array of `StrapiContent` and returns an array of React components.

It uses the `__component` string (eg `basic.paragraph`) to select the correct props mapping function (eg `getRichTextProps(StrapiParagraph) -> RichTextProps` and React Component (eg `<RichText>`).

`<PageContent>` currently needs:
- to know about each `__component` string (which should be internal to each strapi component)
- import all front-end React components
- import all prop mapping functions from `cms/models/`

## This PR
Instead of mapping the props, each Strapi Component would now responsible for returning a React component, with whatever props conversion it might need to do. This keeps the scope for each component much more localized and makes `PageContent` a simple React component that imports other React components. I do agree that these StrapiComponents now should't live in `cms/models` but rather `cms/components`.

**Review by frederike-ramin — APPROVED** [review]
Thanks for taking another look at this! I'm still not 100% convinced that this is the best way to do this (because we tie two "far away" parts of the app very tightly together), but at least it's consistent and we can have a conversation on how to improve this at a later point. So please feel free to merge. :)
